### DISCUSSION TITLE
Traffic of IPv4 in IPv6 with virtual IP is not appearing on vti interface

### DISCUSSION THREAD

**Discussion by SvenRoederer** [strongswan/strongswan#1709] [General]
I have a working setup (Strongswan 5.9.5) of IPv4 roadwarrior-splittunnel using a virtual IP (RFC1918) on a VTI interface.
These days the ISP assigned an IPv6-net, which I added to the current connection.The client connects (NCP Entry Client) via IPv6 and tunnel get established.
In contrast to IPv4 the transported traffic appears on the external interface but not on the existing vti interface.

My assumption is that after decryption the traffic appears on the vti-interface, no matter if IPv4 or IPv6 transport is used. 
Is there some additional setting required, to make this setup work?

**Comment by tobiasbrunner** [comment]
> which I added to the current connection

What does "added" mean?

> there some additional setting required, to make this setup work?

Since you haven't provided any details on your current setup (configs, logs, status, basically what's requested in #196), it's difficult to say.

**Comment by SvenRoederer** [comment]
> What does "added" mean?

just added this new IPv6-address as a 2nd local-address of the existing configuration.

~~~
root@vpn-extdev:~# swanctl -L
no files found matching '/var/ipsec/strongswan.conf'
CVT-extdev-vpn: IKEv2, no reauthentication, rekeying every 14400s
  local:  77.245.x.x
  local:  2a03:5c40:x::0
  remote: %any
  local public key authentication:
    id: C=DE, L=Berlin, O=..., CN=sw-dev...
    certs: C=DE, L=Berlin, O=..., CN=sw-dev...
  remote EAP_TLS authentication:
    eap_id: %any
    cacerts: C=DE, L=Berlin, O=...., CN=VPN-CA
  net: TUNNEL, rekeying every 3600s
    local:  172.16.y.y/29
    remote: dynamic

root@vpn-extdev:~# swanctl -A
no files found matching '/var/ipsec/strongswan.conf'
rw_pool              172.16.y.z                        0 / 13 / 13

~~~


VTI interface is setup by builtin OpenWrt-management (`/etc/config/network`)

~~~
config interface 'dialin'
        option proto 'vti'
        option peeraddr '0.0.0.0'
        option ikey '42'
        option okey '42'
        option tunlink 'wan'

config interface 'dialin_static'
        option proto 'static'
        option ipaddr '172.16.y.y/28'
        option device '@dialin'
~~~

results in 

~~~
root@vpn-extdev:~# ip addr
19: ip_vti0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
20: vti-dialin@br-wan: <NOARP,UP,LOWER_UP> mtu 1280 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ipip 77.245.x.x brd 0.0.0.0
    inet 172.16.y.y/28 brd 172.16.y.y scope global vti-dialin
       valid_lft forever preferred_lft forever
    inet6 fe80::200:5efe:4df5:2433/64 scope link
       valid_lft forever preferred_lft forever

root@vpn-extdev:~# ip tunnel
ip_vti0: ip/ip remote any local any ttl inherit nopmtudisc key 0
vti-dialin: ip/ip remote any local 77.245.x.x dev br-wan ttl inherit nopmtudisc key 42


root@vpn-extdev:~# ip route show table all
172.16.y.65 dev vti-dialin table vpn scope link               <--- a client-address of rw_pool
172.16.y.75 dev vti-dialin table vpn scope link               <--- a client-address of rw_pool
default via 77.245.x.x dev br-wan proto static
77.245.x.x/29 dev br-wan proto kernel scope link src 77.245.x.x2
172.16.y.64/28 dev vti-dialin proto kernel scope link src 172.16.y.78
local 77.245.x.x dev br-wan table local proto kernel scope host src 77.245.x.x
local 127.0.0.0/8 dev lo table local proto kernel scope host src 127.0.0.1
local 127.0.0.1 dev lo table local proto kernel scope host src 127.0.0.1
local 172.16.y.78 dev vti-dialin table local proto kernel scope host src 172.16.y.78
fd7a:44d5:8432::/64 dev wan proto static metric 1024 linkdown pref medium
unreachable fd7a:44d5:8432::/48 dev lo proto static metric 2147483647 pref medium
default via 2a03:5c40:8:0:ffff:ffff:ffff:ffff dev br-wan proto static metric 1024 pref medium
local ::1 dev lo table local proto kernel metric 0 pref medium
{... removed the lines of anycast-, multicast-, linklocal-routes and unrelated interfaces]
~~~

All files below /etc/strongswan are default (and don't contain any explicit settings beside `load = yes` of the installed plugins. My local change is in `/etc/strongswan.d/vti.conf` only

~~~
root@vpn-extdev:~# cat /etc/strongswan.d/vti.conf
# set options for vti
# this file will be inclueded after regular-files have been parsed,
# so this settings win

charon {
        install_virtual_ip = no
        install_routes = no
}
~~~

the "updown-scipt" of swantctl-children section

~~~
#!/bin/sh

logger -t IpSec-VPN "starting up"

# set "charon.install_virtual_ip = no" to prevent the daemon from also installing the Virtual-IP

set -o nounset
set -o errexit

set >/tmp/vti-env.txt

VTI_IF="vti-dialin"
TABLE=vpn

if [ -z "$(grep ${TABLE} /etc/iproute2/rt_tables)" ]; then
        logger -t IpSec-VPN "named routing-table ${TABLE} not defined!"
        exit 1
fi
logger -t IpSec-VPN "startup-tests passed"


case "${PLUTO_VERB}" in
    up-client)
        VTI_IF="vti-dialin"
        logger -t IpSec-VPN "client connected: ${PLUTO_XAUTH_ID} using address ${PLUTO_PEER_SOURCEIP}"
        logger -t ipsec-vti "setting default-route for lan-clients via dev ${VTI_IF}"
        ip route add ${PLUTO_PEER_CLIENT} dev "${VTI_IF}" table ${TABLE} || /bin/true
        sysctl -w "net.ipv4.conf.${VTI_IF}.disable_policy=1"
        ;;
    down-client)
        logger -t IpSec-VPN "client disconnected: ${PLUTO_XAUTH_ID} / ${PLUTO_PEER_CLIENT}"
#        ip tunnel del "${VTI_IF}"
        ;;
esac

~~~

> In contrast to IPv4 the transported traffic appears on the external interface but not on the existing vti interface.

`br-wan` = external interface
`vti-dialin`= vti interface the decrypted clients-traffic gets routed in/out

**Reply by tobiasbrunner** [reply]
> VTI interface is setup by builtin OpenWrt-management

Note that this VTI interface is IPv4-only and bound to the IPv4 address (each interface can only handle a single address family, one of the drawbacks [compared to XFRM interfaces](https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html#_xfrm_interfaces_on_linux)).

**Reply by SvenRoederer** [reply]
So for VTI I will need a VTIv6, in addition to the existing VTIv4?
But as it seems VTIv6 can only handle IPv6 packets. So decapsulating the IPv4 payload of the IPv6-tunnel will require settingup another IPv4inIPv6-tunnelung interface, right?

**Reply by tobiasbrunner** [reply]
I think it's only a problem with the transport addresses (i.e. the addresses of the SAs). So if those can be IPv4 or IPv6, you need two interfaces if you share them (you could also create a separate interface for every connection, I guess). The address family of the traffic sent over the interface/SA shouldn't matter.

**Comment by SvenRoederer** [comment]
I spend some time in trying different ways, but most easy was to switch to XFRM. This avoids any duplication of configuration (IpSec-connection or Interfaces). In my dual-stack tunnel (Ipv4 + IPv6 clients that connecting) where only IPv4-traffic is send over the tunnel.

changes just involved 
* changing the "virtual interface" from VTI to XFRM
* adding `if_id_in` and `if_id_out` to the swanctl configration (inplace of `mark_in`and `mark_out`)
* the `sysctl -w "net.ipv4.conf.${VTI_IF}.disable_policy=1"` call is not needed anymore  inside the updown-script
* the `install_routes" option don need to be changed

So the strongswan configuration became more simple and the network routing is also straight forward no matter of protocol-family.


@tobiasbrunner  Thanks for pointing me to the right direction.
